src/Crypto.ts

```diff
@@ -1,150 +1,152 @@
+  //////////////
+  decryptChunkSync(chunkBuffer: Buffer): Buffer {
+    // Deconstruct chunk into metadata and encrypted data
+    const { salt, initVector, authTag, encryptedBuffer } = deconstructChunk(chunkBuffer)
 
-    return
-  }
+    // Reset decipher
+    this.resetSync(this.masterKey, initVector, salt, authTag)
 
-  private async resetDecipher(initVector: Buffer) {
-    this.initVector = initVector
-    if (this.useWebWorkers && this.cryptorWorker) {
-      return await this.cryptorWorker.init(this.algorithm, this.key, initVector)
-    } else {
-      this.decipher = crypto.createDecipheriv(this.algorithm, this.key, this.initVector)
-    }
-    return
-  }
+    // Decrypt into blockBuffer
+    const blockBuffer = Buffer.concat([this.decipher.update(encryptedBuffer), this.decipher.final()])
 
-  getRandomInitVectorSync(): Buffer {
-    return Buffer.from(crypto.randomBytes(this.initVector.length))
+    return blockBuffer
   }
 
-  async getRandomInitVector(): Promise<Buffer> {
-    return Buffer.from(crypto.randomBytes(this.initVector.length))
-  }
+  async decryptChunk(chunkBuffer: Buffer): Promise<Buffer> {
+    if (this.useWebWorkers) {
+      if (!(await this.waitForCryptoWorkerInit())) {
+        throw(Error('CryptoWorker does not exist'))
+      }
+      // Decrypt into blockBuffer
+      return Buffer.from(await this.cryptoWorker!.decryptChunk(chunkBuffer, this.masterKey, this.algorithm))
+    } else {
+      // Deconstruct chunk into metadata and encrypted data
+      const { salt, initVector, authTag, encryptedBuffer } = deconstructChunk(chunkBuffer)
 
-  private pbkdfSync(pass: string | Buffer, salt = '', algo = 'sha256', keyLen = 32, numIterations = 10000): Buffer {
-    return Buffer.from(crypto.pbkdf2Sync(pass, salt, numIterations, keyLen, algo))
-  }
+      // Reset decipher
+      this.resetSync(this.masterKey, initVector, salt, authTag)
 
-  private async pbkdf(pass: string | Buffer, salt = '', algo = 'sha256', keyLen = 32, numIterations = 10000, callback: (err: Error | null, key: Buffer) => void) {
-    crypto.pbkdf2(pass, salt, numIterations, keyLen, algo, (err, key) => {
-      callback(err, Buffer.from(key))
-    })
+      // Decrypt into blockBuffer
+      return Buffer.concat([this.decipher.update(encryptedBuffer), this.decipher.final()])
+    }
   }
 
+  // ========= Convenience functions ============= //
   hashSync(data: string | Buffer, outputEncoding: 'hex' | 'latin1' | 'base64' = 'hex'): Buffer {
-    const hash = crypto.createHash('sha256')
+    const hash = this.cryptoLib.createHash('sha256')
     hash.update(data)
-    return Buffer.from(hash.digest())
+    return hash.digest()
+  }
+
+  private async delay(ms: number) {
+    return new Promise( resolve => setTimeout(resolve, ms) )
+  }
+  private async waitForCryptoWorkerInit(): Promise<boolean> {
+    for (let trial=0; trial < 10; trial++) {
+      if (this.cryptoWorker) {
+        return true
+      } else {
+        await this.delay(100)
```

Comment:
EFS doesn't concern itself with the OS problems. And not being able to start a cryptoworker sounds an OS problem. And so here we should expect that the worker does start. We can have a configurable timeout (that's a parameter that can be set by the user when they construct EFS). And if this timeout passes by, then we can throw an exception. The timeout can be set by default to `null` which would mean that we just wait forever.

Why doesn't threads.js give us a "promise" based interface to starting a worker? `startWorker.then(() => ...)`.

One way to deal with this is to make the constructor block until it has constructed the worker. And this seems to make sense, because we wouldn't start encrypting anything until we have a worker ready. We can use a white loop to do this. But we should try to have a "sleep".

One issue with this, is that in Polykey we expect to have many vaults and each vault is backed by EFS. In this current implementation, we would have a webworker per vault. Doing this only makes sense when the process has some process stateful context that matters to that process. Our webworker process is just doing pure computation. And we pass everything it needs when it does it. Therefore it seems like we should have the ability to share a webworker pool across multiple instances of EFS.

To do this, we can abstract the "initialisation" of workers to a separate utils file. And let js-polykey call that ahead of time to create the workers (and then in this case, it can actually do it asynchronously). And then chain that with the code that constructs the EFS. This means one can pass `worker_pool` (and can consist of just 1 worker) to the EFS constructor and make it use it. If no worker pool is passed, we don't use workers at all.

Reply:
So it turns out the worker pool initialisation is synchronous! so none of this delay stuff. So I think a util function exposed to the user that initialises the pool first is redundant and we can just leave it in there. But I did it anyway because as you said, multiple EFS instances will want to share the same worker pool.